HTML-JS-PY keywords (#825)
Codecademy/docs: content/python/concepts/classes/classes.md

## Creating a Class

In Python, classes are defined using the [`class`](https://www.codecademy.com/resources/docs/python/keywords/class) keyword.

```py
class Home:
  # Class body starts here
```

**Note**: The example above would actually be invalid because class definitions cannot be empty. However, the `pass` statement can be used as a placeholder to avoid errors:

```py
class Home:
  pass
```

In the class body, the `self` keyword, followed by a period `.`, can be used to call its methods and access its variables, as shown in the `__init__()` [dunder method](https://www.codecademy.com/resources/docs/python/dunder-methods) below:

```py
class Home:
  def __init__(self, rooms, stories):
    # Setting instance variables
    self.rooms = rooms
    self.stories = stories
```

### Class Instances

> Objects can be created or instantiated, from classes. These objects are known as class instances and are created by setting a variable equal to the class name followed by parentheses `()`:

Reviewer: ```suggestion
Objects can be created or instantiated from classes. These objects are known as class instances and are created by setting a variable equal to the class name followed by parentheses `()`:
```

```py
my_home = Home()
```

> Here, the instance name is `my_home`, and the class is `Home`. Calling this line implicitly calls the `Home` class's [`__init__()`](https://www.codecademy.com/resources/docs/python/dunder-methods/__init__) method.

Reviewer: ```suggestion
Here, the instance name is `my_home`, which derives from the `Home` class. Calling this line implicitly calls the `Home` class's [`__init__()`](https://www.codecademy.com/resources/docs/python/dunder-methods/__init__) method.
```

  - 'Data Science'
  - 'Web Development'
Tags:
  - 'Attributes'
  - 'Classes'
  - 'Methods'
  - 'Objects'
  - 'OOP'
CatalogContent:
  - 'learn-python-3'
  - 'paths/computer-science'
---

**Classes** are templates used to define the properties and methods of objects in code. They can describe the kinds of data that the class holds, and also how a programmer interacts with that data.

The usage of classes are a key element of [object-oriented programming](https://www.codecademy.com/resources/docs/general/object-oriented-programming) (OOP) where "class instances" are created to implement design principles such as [inheritance](https://www.codecademy.com/resources/docs/general/inheritance) and [encapsulation](https://www.codecademy.com/resources/docs/general/encapsulation).

## Creating a Class

```py
class Home:
  # Class body starts here
```

> In Python, classes are defined using the [`class`](https://www.codecademy.com/resources/docs/python/keywords/class) keyword. Within the code of the class, we can call its methods and access its variables. We do this by using `self` followed by a period `.` and then followed by the method or variable.

Reviewer 2: We need an example of this here.

### Class Instances

Objects can be created or instantiated, from classes. These objects are known as class instances. To create an instance of a class, we can set a variable equal to the class name followed by parentheses `()`:

```py
my_home = Home()
```

Here, the instance name is `my_home`, and the class is `Home`.

### Attributes

> Class attributes are variables that are defined outside of all methods and have the same value for every instance of the class.

Reviewer 2: ```suggestion
Class attributes are variables that are defined outside of all methods and have the same value for every instance of the class. They also can be accessed via the class name rather than the instance name. Setting the variable via the class name will change it for all instances. **Note:** setting it via an instance name will break the connection for that instance.
```

Reviewer 2: You don't have to use this language, but some discussion of the idiosyncrasies of python class attributes is warranted.

```py
class Bird:
  # Class attribute
  leg_count = 2

parakeet = Bird()
parrot = Bird()

print(parakeet.leg_count) # Output: 2
print(parrot.leg_count)   # Output: 2

> ```

Reviewer 2: It might be a good idea to have an example where the Class attribute is changed, and we show it changing for the various instances.

> blue_home.paint_wall("blue")

Reviewer 2: It might be good to provide something for the method body above (setting an attribute perhaps.)

Author: I ended up ditching the `Home` class/`.paint_wall()` example and replaced with a `feed_bird()` method inside the `Bird` class, instead.

```

#### Static Methods

It is also possible to create method that only apply to the class itself, not instances of the class. These can be distinguised with `@classmethod` and `@staticmethod`:

```py
class Home:
  name="Code Ninja"
  rooms = 4
  stories = 2

  @staticmethod
  def is_on_market(home):
    if(home.name == ""):
      return True
    else:
      return False

  @classmethod
  def paint_wall(self, color):
    return f"Painting wall with the color {color}."

home = Home()

print(Home.is_on_market(home))
# Output: False
```

> #### Customized Class Methods

Reviewer 2: It might be worth it to expand this to cover more dunder/magic methods.

Author: This actually reminds me of something I was considering. Should we do one of the following?

- Create a new `terms` folder under `concepts/classes` with entries for these methods. (e.g., `__init__.md`, `__repr__.md`, `__str__.md`)
- Create a new _concept entry_ (maybe it could be called `Dunder Methods`?) and then create new term entries under there?

Reviewer 2: @Dusch4593, My vote would be for the latter. A new entry called "Dunder Methods".

  def working(self, employee_name):
    self.name = employee_name
    print(f"{employee_name} is working")
```

Afterward, the `Employee` class can be imported into other files where new instances and methods can be created. This makes the code efficient, reusable, and maintainable:

```py
# other_file.py
from employee import Employee

def create_employee():

>   print("Employee is starting thier job...")

Reviewer: ```suggestion
  print("Employee is starting their job...")
```

  employee1 = Employee()
  employee1.name = "Blake"
  employee1.working(employee1.name)

create_employee();
```

Running the code in `other_file.py` will output the following:

```shell

> Employee is starting thier job...

Reviewer: ```suggestion
Employee is starting their job...
```

Blake is working
```

## Codebyte Example

Defined methods and variables can be accessed within the class definition. This is done with the `self` keyword, followed by a period `.`, and then followed by the method or variable. In the example below, the defined `.getName()` method is used inside the `.sayHi()` method:

> ```py

Reviewer: ```suggestion
```codebyte/python
```
